File: java/src/main/java/io/github/muthuishere/toolnexus/HttpTool.java
package io.github.muthuishere.toolnexus;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.HashSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * HTTP / REST tool (source {@code "http"}). Declares a remote endpoint as a
 * uniform {@link Tool}. Mirrors the JS reference ({@code js/src/http.ts}).
 *
 * <ul>
 *   <li>{@code {placeholder}} URL substitution (consumed from args).</li>
 *   <li>{@code query} arg names → querystring (GET → all args to query).</li>
 *   <li>body json|form|raw for non-GET requests.</li>
 *   <li>{@code ${ENV}} header expansion from {@code System.getenv} (never logged).</li>
 *   <li>non-2xx ⇒ {@code "HTTP <status>: <body>"} error; 2xx ⇒ body + metadata{status}.</li>
 * </ul>
 */
public final class HttpTool implements Tool {
    private static final long DEFAULT_TIMEOUT = 30_000L;
    private static final Pattern PLACEHOLDER = Pattern.compile("\\{(\\w+)}");
    private static final HttpClient CLIENT = HttpClient.newHttpClient();

    public static final class Options {
        public String name;
        public String description;
        public String method;
        public String url;
        public Map<String, String> headers;
        public List<String> query;
        public String body; // "json" | "form" | "raw"
        public Map<String, Object> inputSchema;
        public Long timeout;
        public String resultMode; // "text" | "json" | "status+text"
    }

    private final Options opts;
    private final String method;
    private final Set<String> querySet;
    private final Map<String, Object> inputSchema;

    private HttpTool(Options opts) {
        this.opts = opts;
        this.method = opts.method.toUpperCase();
        this.querySet = new HashSet<>(opts.query == null ? List.of() : opts.query);
        this.inputSchema = opts.inputSchema != null ? opts.inputSchema : emptySchema();
    }

    public static HttpTool of(Options opts) {
        return new HttpTool(opts);
    }

    @Override
    public String name() {
        return opts.name;
    }

    @Override
    public String description() {
        return opts.description;
    }

    @Override
    public Map<String, Object> inputSchema() {
        return inputSchema;
    }

    @Override
    public String source() {
        return "http";
    }

    @Override
    public ToolResult execute(Map<String, Object> args, ToolContext ctx) {
        Map<String, Object> a = new LinkedHashMap<>(args == null ? Map.of() : args);

        // 1. substitute {placeholders} in the URL from args (consumed afterwards)
        StringBuffer sb = new StringBuffer();
        Matcher m = PLACEHOLDER.matcher(opts.url);
        while (m.find()) {
            String key = m.group(1);
            Object val = a.remove(key);
            String enc = URLEncoder.encode(val == null ? "" : String.valueOf(val), StandardCharsets.UTF_8);
            m.appendReplacement(sb, Matcher.quoteReplacement(enc));
        }
        m.appendTail(sb);
        String url = sb.toString();

        // 2. querystring args
        List<String> qsParts = new ArrayList<>();
        List<String> keys = new ArrayList<>(a.keySet());
        for (String key : keys) {
            if (querySet.contains(key) || method.equals("GET")) {
                Object v = a.remove(key);
                qsParts.add(URLEncoder.encode(key, StandardCharsets.UTF_8) + "="
                        + URLEncoder.encode(String.valueOf(v), StandardCharsets.UTF_8));
            }
        }
        if (!qsParts.isEmpty()) {
            url += (url.contains("?") ? "&" : "?") + String.join("&", qsParts);
        }

        // 3. body
        Map<String, String> expanded = McpSource.expandEnvHeaders(opts.headers);
        Map<String, String> headers = expanded == null ? new LinkedHashMap<>() : new LinkedHashMap<>(expanded);
        String bodyInit = null;
        if (!method.equals("GET") && !method.equals("HEAD") && !a.isEmpty()) {
            String mode = opts.body == null ? "json" : opts.body;
            if (mode.equals("json")) {
                headers.putIfAbsent("Content-Type", "application/json");
                bodyInit = Json.stringify(a);
            } else if (mode.equals("form")) {
                headers.putIfAbsent("Content-Type", "application/x-www-form-urlencoded");
                List<String> formParts = new ArrayList<>();
                for (Map.Entry<String, Object> e : a.entrySet()) {
                    formParts.add(URLEncoder.encode(e.getKey(), StandardCharsets.UTF_8) + "="
                            + URLEncoder.encode(String.valueOf(e.getValue()), StandardCharsets.UTF_8));
                }
                bodyInit = String.join("&", formParts);
            } else {
                Object b = a.get("body");
                bodyInit = b == null ? "" : String.valueOf(b);
            }
        }

        long timeout = ctx != null && ctx.timeoutMs() != null
                ? ctx.timeoutMs()
                : (opts.timeout != null ? opts.timeout : DEFAULT_TIMEOUT);

        HttpRequest.Builder rb = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofMillis(timeout));
        for (Map.Entry<String, String> e : headers.entrySet()) {
            rb.header(e.getKey(), e.getValue());
        }
        HttpRequest.BodyPublisher publisher = bodyInit == null
                ? HttpRequest.BodyPublishers.noBody()
                : HttpRequest.BodyPublishers.ofString(bodyInit, StandardCharsets.UTF_8);
        rb.method(method, publisher);

        try {
            HttpResponse<String> res = CLIENT.send(rb.build(), HttpResponse.BodyHandlers.ofString());
            int status = res.statusCode();
            String text = res.body();
            Map<String, Object> meta = new LinkedHashMap<>();
            meta.put("status", status);
            if (status < 200 || status >= 300) {
                return new ToolResult("HTTP " + status + ": " + text, true, meta);
            }
            String output;
            if ("status+text".equals(opts.resultMode)) {
                output = status + "\n" + text;
            } else if ("json".equals(opts.resultMode)) {
                output = Json.stringify(Json.parseLoose(text));
            } else {
                output = text;
            }
            return new ToolResult(output, false, meta);
        } catch (Exception e) {
            return ToolResult.error(e.getMessage() == null ? String.valueOf(e) : e.getMessage());
        }
    }

    private static Map<String, Object> emptySchema() {
        Map<String, Object> s = new LinkedHashMap<>();
        s.put("type", "object");
        s.put("properties", new LinkedHashMap<>());
        s.put("additionalProperties", false);
        return s;
    }
}

File: csharp/src/Toolnexus/HttpTool.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Toolnexus;

/// <summary>
/// HTTP / REST tool (source <c>"http"</c>). Declares a remote endpoint as a uniform
/// <see cref="ITool"/>. Mirrors the JS reference (<c>js/src/http.ts</c>).
/// <list type="bullet">
///   <item><c>{placeholder}</c> URL substitution (consumed from args).</item>
///   <item><c>Query</c> arg names → querystring (GET → all args to query).</item>
///   <item>body json|form|raw for non-GET requests.</item>
///   <item><c>${ENV}</c> header expansion from the environment (never logged).</item>
///   <item>non-2xx ⇒ <c>"HTTP &lt;status&gt;: &lt;body&gt;"</c> error; 2xx ⇒ body + metadata{status}.</item>
/// </list>
/// </summary>
public sealed partial class HttpTool : ITool
{
    private const long DefaultTimeout = 30_000L;
    private static readonly HttpClient Client = new();

    [GeneratedRegex(@"\{(\w+)\}")]
    private static partial Regex Placeholder();

    public sealed class Options
    {
        public string Name { get; set; } = "";
        public string Description { get; set; } = "";
        public string Method { get; set; } = "GET";
        public string Url { get; set; } = "";
        public IDictionary<string, string>? Headers { get; set; }
        public List<string>? Query { get; set; }
        public string? Body { get; set; } // "json" | "form" | "raw"
        public IDictionary<string, object?>? InputSchema { get; set; }
        public long? Timeout { get; set; }
        public string? ResultMode { get; set; } // "text" | "json" | "status+text"
    }

    private readonly Options _opts;
    private readonly string _method;
    private readonly HashSet<string> _querySet;

    public string Name => _opts.Name;
    public string Description => _opts.Description;
    public IDictionary<string, object?> InputSchema { get; }
    public string Source => "http";

    private HttpTool(Options opts)
    {
        _opts = opts;
        _method = opts.Method.ToUpperInvariant();
        _querySet = new HashSet<string>(opts.Query ?? new List<string>());
        InputSchema = opts.InputSchema ?? NativeTool.EmptySchema();
    }

    public static HttpTool Of(Options opts) => new(opts);

    public async Task<ToolResult> ExecuteAsync(IDictionary<string, object?> args, ToolContext? ctx = null)
    {
        var a = new Dictionary<string, object?>(args ?? new Dictionary<string, object?>());

        // 1. substitute {placeholders} in the URL from args (consumed afterwards)
        var url = Placeholder().Replace(_opts.Url, m =>
        {
            var key = m.Groups[1].Value;
            a.Remove(key, out var val);
            return Uri.EscapeDataString(val?.ToString() ?? "");
        });

        // 2. querystring args
        var qsParts = new List<string>();
        foreach (var key in a.Keys.ToList())
        {
            if (_querySet.Contains(key) || _method == "GET")
            {
                a.Remove(key, out var v);
                qsParts.Add(Uri.EscapeDataString(key) + "=" + Uri.EscapeDataString(v?.ToString() ?? ""));
            }
        }
        if (qsParts.Count > 0)
            url += (url.Contains('?') ? "&" : "?") + string.Join("&", qsParts);

        // 3. body
        var headers = McpSource.ExpandEnvHeaders(_opts.Headers) ?? new Dictionary<string, string>();
        string? bodyInit = null;
        string? contentType = null;
        if (_method != "GET" && _method != "HEAD" && a.Count > 0)
        {
            var mode = _opts.Body ?? "json";
            if (mode == "json")
            {
                contentType = "application/json";
                bodyInit = Json.Stringify(a);
            }
            else if (mode == "form")
            {
                contentType = "application/x-www-form-urlencoded";
                bodyInit = string.Join("&", a.Select(e =>
                    Uri.EscapeDataString(e.Key) + "=" + Uri.EscapeDataString(e.Value?.ToString() ?? "")));
            }
            else
            {
                bodyInit = a.TryGetValue("body", out var b) ? b?.ToString() ?? "" : "";
            }
        }

        var timeoutMs = ctx?.TimeoutMs ?? _opts.Timeout ?? DefaultTimeout;

        try
        {
            using var req = new HttpRequestMessage(new HttpMethod(_method), url);
            if (bodyInit != null)
            {
                req.Content = new StringContent(bodyInit, Encoding.UTF8);
                if (contentType != null)
                    req.Content.Headers.ContentType =
                        new System.Net.Http.Headers.MediaTypeHeaderValue(contentType) { CharSet = null };
            }
            foreach (var (k, v) in headers)
            {
                if (!req.Headers.TryAddWithoutValidation(k, v) && req.Content != null)
                    req.Content.Headers.TryAddWithoutValidation(k, v);
            }

            using var cts = CancellationTokenSource.CreateLinkedTokenSource(ctx?.CancellationToken ?? default);
            cts.CancelAfter(TimeSpan.FromMilliseconds(timeoutMs));

            using var res = await Client.SendAsync(req, cts.Token).ConfigureAwait(false);
            var status = (int)res.StatusCode;
            var text = await res.Content.ReadAsStringAsync(cts.Token).ConfigureAwait(false);
            var meta = new Dictionary<string, object?> { ["status"] = status };

            if (status < 200 || status >= 300)
                return new ToolResult($"HTTP {status}: {text}", true, meta);

            string output;
            if (_opts.ResultMode == "status+text")
                output = status + "\n" + text;
            else if (_opts.ResultMode == "json")
                output = Json.Stringify(Json.ParseLoose(text));
            else
                output = text;

            return new ToolResult(output, false, meta);
        }
        catch (Exception e)
        {
            return ToolResult.Error(e.Message ?? e.ToString());
        }
    }
}
